### PR TITLE
[cli] Add a replay-events debug command.

### PR DESCRIPTION
**Pull request by pgavlin**
This command renders a recorded stream of engine events using either the
progress or diff renderer. This is useful for debugging problems with
these renderers.

### CODE REVIEW

File: pkg/backend/httpstate/state.go [hunk 1]
```diff
@@ -168,7 +168,7 @@ func (u *cloudUpdate) recordEngineEvents(startingSeqNumber int, events []engine.
 
 	var apiEvents apitype.EngineEventBatch
 	for idx, event := range events {
-		apiEvent, convErr := display.ConvertEngineEvent(event)
+		apiEvent, convErr := display.ConvertEngineEvent(event, false)
```
Comment (Frassle) [inline pkg/backend/httpstate/state.go:171]:
```suggestion
		apiEvent, convErr := display.ConvertEngineEvent(event, false /* showSecrets */)
```

File: pkg/backend/display/display.go [hunk 2]
```diff
@@ -56,21 +56,21 @@ func ShowEvents(
...
-		ShowWatchEvents(op, action, events, done, opts)
+		ShowWatchEvents(op, events, done, opts)
 	default:
 		contract.Failf("Unknown display type %d", opts.Type)
 	}
 }
 
 func logJSONEvent(encoder *json.Encoder, event engine.Event, opts Options, seq int) error {
-	apiEvent, err := ConvertEngineEvent(event)
+	apiEvent, err := ConvertEngineEvent(event, true)
```
Comment (Frassle) [inline pkg/backend/display/display.go:73]:
```suggestion
	apiEvent, err := ConvertEngineEvent(event, true /* showSecrets */)
```

File: pkg/cmd/pulumi/replay_events.go [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+			for _, e := range events {
+				eventChannel <- e
+			}
+			<-doneChannel
+
+			return nil
+		}),
+	}
+
+	cmd.PersistentFlags().BoolVarP(
```
Comment (Frassle) [inline pkg/cmd/pulumi/replay_events.go:120]: Its a bit sad this is needed but looks to be a gnarly untangling to get rid of it.

Comment (pgavlin) [reply pkg/cmd/pulumi/replay_events.go:120]: Yeah. We might be able to scrape this out of the events if we have enough of them--I think that the resource events include a bit that indicates whether or not they're part of a preview. Feels like a better solution would be to issue a "start" event that contains this info, though.

**Comment by pgavlin** [comment]
> > These changes also switch the JSON event log to show secrets
>
> I am very open to making this configurable--event logs are still useful without rich secrets, they're just lower-fidelity. In the case that motivated this change, I needed high-fidelity secrets in order to debug our diff rendering logic.

Curious if anybody has thoughts on this?

**Comment by iwahbe** [comment]
> > > These changes also switch the JSON event log to show secrets
> > 
> > 
> > I am very open to making this configurable--event logs are still useful without rich secrets, they're just lower-fidelity. In the case that motivated this change, I needed high-fidelity secrets in order to debug our diff rendering logic.
> 
> Curious if anybody has thoughts on this?

I think this is fine. Since this is a debug command, we can presume users know that this will expose secrets. That being said, making the behavior configurable would expand the set of users we could ask to run this command, and should add only marginal complexity.

**Comment by pgavlin** [comment]
> > > > These changes also switch the JSON event log to show secrets
> > > 
> > > 
> > > I am very open to making this configurable--event logs are still useful without rich secrets, they're just lower-fidelity. In the case that motivated this change, I needed high-fidelity secrets in order to debug our diff rendering logic.
> > 
> > 
> > Curious if anybody has thoughts on this?
> 
> I think this is fine. Since this is a debug command, we can presume users know that this will expose secrets.[^1] That being said, making the behavior configurable would expand the set of users we could ask to run this command, and should add only marginal complexity.
> 
> [1]: If this assumption is invalid, we should definitely put exposing secrets behind a flag.

I ended up reverting this aspect of the change. Retaining plaintext ends up being pretty much useless because the plaintext is removed upstream by the engine.

**Review by Frassle — APPROVED** [review]

**Pull request merged**